### PR TITLE
Add internal n/w bridge to firewalld docker zone

### LINKED ISSUE #47229: Containers cannot communicate with each other when using an internal network
### Description

```
version: "3.8"

networks:
  database:
    driver: bridge
    internal: true
    ipam:
      driver: default
      config:
      - subnet: 172.16.240.0/24

services:
  nginx:
    image: nginx
    container_name: nginx
    networks:
      #frontend:
      #backend:
      database:

  debug:
    image: curlimages/curl
    container_name: debug
    command: "sh"
    stdin_open: true
    tty: true
    networks:
      #frontend:
      #backend:
      database:
```
I tested on openSUSE and fedora systems. On the openSUSE system, the ping goes through but not the port, and on the Fedora system it doesn't even ping through.

### Reproduce

docker compose up -d
docker exec debug ping nginx
docker exec debug curl -Iv nginx
> connect to 172.16.240.2 port 80 from 172.16.240.3 port 56700 failed: Operation timed out
> * Failed to connect to nginx port 80 after 132845 ms: Couldn't connect to server

### Expected behavior

docker exec debug curl -Iv nginx
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 172.16.240.2:80...
* Connected to nginx (172.16.240.2) port 80 (#0)
> HEAD / HTTP/1.1
> Host: nginx
> User-Agent: curl/7.88.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: nginx/1.25.3
< Date: Fri, 26 Jan 2024 03:49:26 GMT
< Content-Type: text/html
< Content-Length: 615
< Last-Modified: Tue, 24 Oct 2023 13:46:47 GMT
< Connection: keep-alive
< ETag: "6537cac7-267"
< Accept-Ranges: bytes
< 
  0   615    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Connection #0 to host nginx left intact
HTTP/1.1 200 OK
Server: nginx/1.25.3
Date: Fri, 26 Jan 2024 03:49:26 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Tue, 24 Oct 2023 13:46:47 GMT
Connection: keep-alive
ETag: "6537cac7-267"
Accept-Ranges: bytes
```

### docker version

```bash
Client: Docker Engine - Community
 Version:           25.0.1
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        29cf629
 Built:             Tue Jan 23 23:11:29 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          25.0.1
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       71fa3ab
  Built:            Tue Jan 23 23:09:46 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.27
  GitCommit:        a1496014c916f9e62104b33d1bb5bd03b0858e59
 runc:
  Version:          1.1.11
  GitCommit:        v1.1.11-0-g4bccb38
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    25.0.1
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.2
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 2
 Server Version: 25.0.1
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: a1496014c916f9e62104b33d1bb5bd03b0858e59
 runc version: v1.1.11-0-g4bccb38
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.5.6-300.fc39.x86_64
 Operating System: Fedora Linux 39 (Server Edition)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 3.809GiB
 Name: localhost.localdomain
 ID: 964f3733-ef63-4107-a6e5-df4333d1894a
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by robmry**
Fixes https://github.com/moby/moby/issues/47229

**- What I did**

Containers attached to an 'internal' bridge network are unable to communicate when the host is running firewalld.

Non-internal bridges are added to a trusted 'docker' firewalld zone, but internal bridges were not - they are now.

**- How I did it**

For a non-internal network, the Add/Del functions are called by `iptables.ProgramChain` (multiple times per bridge). There's there's no equivalent in the iptables package for an internal network. So, this change calls them from the bridge driver, in the function that's used for create and cleanup of an internal bridge's rules.

**- How to verify it**

Tested a local build on openSUSE, using the example in the original ticket.

Without this change, pings and curls over the internal network fail, they hit a 'reject with icmpx admin-prohibited' firewalld rule.

With the change, communication between containers works. The DOCKER-ISOLATION iptables rules are still configured for an internal network, they block traffic to/from addresses outside the network's subnet (checked by looking at counters in `iptables -nvL -t filter`). The bridge still does not get an IP address, so there's no route for communication with the host. So, it is still an internal network.

_It should be covered by existing tests in 'integration/networking/bridge_test.go' ... but I don't think we run CI on any hosts that use firewalld?_

**- Description for the changelog**

Include 'internal' network bridges in the 'docker' zone for firewalld, so that containers can communicate.

### CODE REVIEW

**Review by corhere — APPROVED** [review]

File: libnetwork/drivers/bridge/setup_ip_tables_linux.go [hunk 1]
```diff
@@ -408,6 +409,17 @@ func setupInternalNetworkRules(bridgeIface string, addr *net.IPNet, icc, insert
 	var version iptables.IPVersion
 	var inDropRule, outDropRule iptRule
 
+	// Either add or remove the interface from the firewalld zone, if firewalld is running.
```
Comment (akerouanton) [inline libnetwork/drivers/bridge/setup_ip_tables_linux.go:412]: It seems `ProgramChain` is only called by the bridge driver. Do you think it'd make sense to move the calls to Add/Del methods out of the `iptables` package to `setupIPTables` (ie. `setupInternalNetworkRules` caller), such that we don't duplicate these in two different places?

Comment (robmry) [reply libnetwork/drivers/bridge/setup_ip_tables_linux.go:412]: Yes, I don't think the firewalld calls belong with the rest of `ProgramChain`, and should be called separately (and once) by the bridge driver.

It looked like a bit of a mess though. And, particularly for firewalld, not well regression tested. So, I figured a minimal change would be more likely to make it into 25.0.3.

And, it's the usual problem with changing an exported package? Maybe that doesn't matter? Or, if it does, we could split `ProgramChain`, and make it call new functions that we use from the bridge driver?

But perhaps that should be a separate refactoring PR?

Comment (robmry) [reply libnetwork/drivers/bridge/setup_ip_tables_linux.go:412]: Excellent! Thanks Cory.

**Review by akerouanton — APPROVED** [review]
LGTM

**Pull request merged**